src/renderer/components/cluster-settings/components/cluster-local-terminal-settings.tsx

```diff
@@ -53,8 +59,45 @@ export class ClusterHomeDirSetting extends React.Component<Props> {
     );
   }
 
-  saveCWD = () => {
-    this.props.cluster.preferences.terminalCWD = this.directory;
+  saveCWD = async () => {
+    if (!this.directory) {
+      this.props.cluster.preferences.terminalCWD = undefined;
+
+      return;
+    }
+
+    try {
+      const dir = resolveTilde(this.directory);
+      const stats = await stat(dir);
+
+      if (stats.isDirectory()) {
+        this.props.cluster.preferences.terminalCWD = dir;
+      } else {
+        Notifications.error(
+          <>
+            <b>Shell Working Directory</b>
+            <p>Provided path is not a directory, your changes were not saved.</p>
+          </>,
+        );
+      }
+    } catch (error) {
+      if (error.code === "ENOENT") {
+        Notifications.error(
+          <>
+            <b>Shell Working Directory</b>
+            <p>Provided path does not exist, your changes were not saved.</p>
+          </>,
+        );
+      } else {
```

Comment:
Shouldn't suppress unknown exceptions here. It hides information, prevents centralized error handling and logging etc.

Reply:
How is this suppressing unknown errors, do you mean I should handle each FS error code separately?

Reply:
I mean that unknown exceptions are unknown for a reason. We should treat them as fatal exceptions which have to be fixed from our side.

So basically, separate fatal exceptions and non fatal exceptions from each other.

- Fatal exception is something unexpected that we need to know about. User has no way to fix it. 
- Non fatal exception is something expected that we don't need to know about. User has to fix it, since it's part of the non-happy case scenario. In this case, e.g. providing path to directory that doesn't exist is "non fatal exception", since user can simply change the directory.  (as you have done)

I can see that you have done changes to the logic handling exceptions, and it's getting better. Most important part for now, is that the unknown exceptions are reported somewhere so that we can start treating them as **bugs to be fixed**. 

In the future, I think that we should start catching the fatal exceptions from top level and using throw as a mechanism since then we have most detailed error stack available when we are trying to fix the issue. But yeah, that's long in the future, since it would mean that we need to get rid of larger scope "try/catch"-usages.

Reply:
We cannot really treat them as fatal since the application should continue to be usable, I would prefer "expected exceptions" vs "unexpected exceptions" (which I know sounds like an oxymoron). 

In this I don't really like the design of JS, and much prefer Rust's design, where errors are just data. It is slightly hard to do that in JS since the entire ecosystem uses `throw`